bots/machine/testvm_core.py

```diff
@@ -0,0 +1,557 @@
+        subprocess.check_call(cmd)
+
+    def download(self, source, dest, relative_dir="."):
+        """Download a file from the test machine.
+        """
+        assert source and dest
+        assert self.ssh_address
+
+        self._ensure_ssh_master()
+        dest = os.path.join(relative_dir, dest)
+
+        cmd = [
+            "scp", "-B",
+            "-P", str(self.ssh_port),
+            "-o", "StrictHostKeyChecking=no",
+            "-o", "UserKnownHostsFile=/dev/null",
+            "-o", "ControlPath=" + self.ssh_master,
+            "-o", "BatchMode=yes",
+            ]
+        if not self.verbose:
+            cmd += ["-q"]
+        cmd += [ "%s@[%s]:%s" % (self.ssh_user, self.ssh_address, source), dest ]
+
+        self.message("Downloading", source)
+        self.message(" ".join(cmd))
+        subprocess.check_call(cmd)
+
+    def download_dir(self, source, dest, relative_dir="."):
+        """Download a directory from the test machine, recursively.
+        """
+        assert source and dest
+        assert self.ssh_address
+
+        self._ensure_ssh_master()
+        dest = os.path.join(relative_dir, dest)
+
+        cmd = [
+            "scp", "-B",
+            "-P", str(self.ssh_port),
+            "-o", "StrictHostKeyChecking=no",
+            "-o", "UserKnownHostsFile=/dev/null",
+            "-o", "ControlPath=" + self.ssh_master,
+            "-o", "BatchMode=yes",
+            "-r",
+          ]
+        if not self.verbose:
+            cmd += ["-q"]
+        cmd += [ "%s@[%s]:%s" % (self.ssh_user, self.ssh_address, source), dest ]
+
+        self.message("Downloading", source)
+        self.message(" ".join(cmd))
+        try:
+            subprocess.check_call(cmd)
+            target = os.path.join(dest, os.path.basename(source))
+            if os.path.exists(target):
+                subprocess.check_call([ "find", target, "-type", "f", "-exec", "chmod", "0644", "{}", ";" ])
+        except:
+            self.message("Error while downloading directory '{0}'".format(source))
+
+    def write(self, dest, content):
+        """Write a file into the test machine
+
+        Arguments:
+            content: Raw data to write to file
+            dest: The file name in the machine to write to
+        """
+        assert dest
+        assert self.ssh_address
+
+        cmd = "cat > '%s'" % dest
+        self.execute(command=cmd, input=content)
+
+    def spawn(self, shell_cmd, log_id):
+        """Spawn a process in the test machine.
+
+        Arguments:
+           shell_cmd: The string to execute by /bin/sh.
+           log_id: The name of the file, realtive to /var/log on the test
+              machine, that will receive stdout and stderr of the command.
+        Returns:
+            The pid of the /bin/sh process that executes the command.
+        """
+        return int(self.execute("{ (%s) >/var/log/%s 2>&1 & }; echo $!" % (shell_cmd, log_id)))
+
+    def journal_cursor(self):
```

Comment:
This and the following two methods don't really conceptually fit into a "SSHConnection" class, they should remain in a "Machine" class.

Reply:
thats true and I understand you PoV, but these methods was interesting to me